Zheng07Sats generalization

aphearin opened the pull request:
This PR gives some further modifications along the lines originally suggested by @mclaughlin6464 in #567 and #568. Summary of changes:

1. (https://github.com/aphearin/halotools/commit/8d378e7176d146a31e291eff3379a157f0ac4305) The Zheng07Sats class now accepts two arguments: `moculate_with_cenocc=False`, and additionally `cenocc_model=None`. This implementation is completely compatible with previous default behavior, but provides the desired abstraction through the cenocc_model keyword. Some new unit-testing explicitly verifies the correct behavior. 

2. (https://github.com/aphearin/halotools/commit/0c163c982189271e99bccca2f7aa825eaabe7e0f) The `PrebuiltHodModelFactory(‘zheng07’)` raises an exception at an attempted use of the cenocc_model keyword. To build a composite model using the generality of the Zheng07Sats model that uses this keyword, it is necessary to use the HodModelFactory, as demonstrated in a new tutorial on advanced usage of the ‘zheng07’ model. Prohibiting the pre-built factory to use this feature is a sort of arbitrary judgement call over what should and should not be considered an entirely new model; hopefully it won't be too much of a hassle to use the real factory directly, but this could be up for discussion. Any strong opinion, @mclaughlin6464?

mclaughlin6464 commented:
Regarding your first point, that seems like the best implementation, better than the PR I made. This way all the legacy code will still work but so-called "power-users" can access this functionality. The one thing I'd like to add is that I think the best possible implementation would be that the satellite object can access whatever the central object is in the overarching model. This makes the most sense as a user, as I think of the model components having awareness of one another. However, doing so is not really possible with the current object hierarchy, so that change would have to be part of a ground-up redesign. Just food for thought. 

For your second point, my concern is with what happens on lines 302 in `zheng07_components`. At this point the parameters from the new `central_occupation_model` are written into the satellite component's `param_dict`, which ultimately makes its way into the `HOD_model_factory`'s `param_dict`. I initially discovered this issue because the parameters for the central occupation model I passed into `HOD_model_factory` were overwritten by the parameters in the generic `Zheng07Cens` object inside the satellite  object. Even now, if the user passes in one central occupation object into `HOD_model_factory` and another object into `Zheng07Sats` they would get undefined behavior, as which parameters end up in the `HOD_model_factory.param_dict` is dependent on the ordering they are accessed via a dictionary iterator. It's hard to imagine how someone could make this mistake, but I've definitely done equally dumb things before :). I think we should add a check into `HOD_model_factory` that ensures either the two objects are equal, or at least their parameters are equal. Do you agree?

aphearin commented:
Thanks for weighing in, @mclaughlin6464. 

Re: point 1. I understand exactly what you mean and agree this is a limitation of the current framework. I hemmed and hawed over this quite a bit during the initial design, and it's the usual thing: nobody at that time needed this (including me), so I just kicked the can down the road and moved on to more pressing features. 

That said, I'm currently finishing up a new feature of the factory that I *think* should permit this generalization, though I'm not 100% sure because I added it in for an entirely different purpose (using subhalo positions for satellite positions). Briefly, the way it works is that a component model will have the power to request any *attribute* of the mock object. Whatever `attrname` is requested gets passed in as a keyword argument whenever one of the `mock_generation_calling_sequence` functions gets called. In my application, the attribute will be a `subhalo_table`, but in your application the attribute will be the `model` object itself. I think that should be sufficient.

aphearin commented:
Re: point 2. At a minimum, I can do something simple to provide further checks on self-consistency. My "solution" in this PR was just to provide explicit documentation pointing out that it is the user responsibility to make a self-consistent choice. Before we go too far down this road, can you think of a realistic use-case in which one would actually *want* to choose one central occupation model for the centrals, but then a *different* central occupation model to modulate the satellites? I cannot think of a use-case that could not be equally well solved by simply writing the additional functionality directly into the satellite model, rather than passing it in. I'm certainly game to work harder on such a feature, but I guess I'd need a bit better understanding of the motivation before tackling it.

aphearin commented:
Also, sorry, but which commit are you referring to for line 302? In the tip of this PR, line 302 of `zheng07_components.py` is the following:

`upper_occupation_bound = float("inf")`.

mclaughlin6464 commented:
And my mistake, I was looking at the master branch. In this PR I'm looking at line 337.

aphearin commented:
Roger. I'll push up the fix for a self-consistency check you suggested shortly, which is a good idea. Then whenever you merge it in this will automatically trigger a Travis re-build on astropy/halotools/master.

aphearin commented:
In https://github.com/mclaughlin6464/halotools/pull/1/commits/ab50e59ec9204d0832ed5705ab069adc900608d0, what I do is search all component models that are `OccupationComponent` instances, and verify that the value bound to *any* `central_occupation_model` attribute is consistent with the *actual* central occupation moment (if one exists at all). The control flow is a little tricky just because the factory permits any number of galaxy types, etc., but I think this enforces the self-consistency that you suggested, and I put in a unit-test to explicitly verify that a minimal problem example would be caught by this implementation.

aphearin commented:
I can keep you in the loop on the other generalization you mentioned here, it should be done within a few days. Even without that I think this is worth going ahead and merging.
